### PR TITLE
WIP Add PathPrefixAdapter, moved prefixing behavior to a trait.

### PR DESCRIPTION
**Pull request by SamMousa**
The test is not done yet.

I'd appreciate some assistance since I'm not familiar with mocking.

I can check that a specific method gets called exactly once, but I'm unsure how to specify that all other functions should not be called.

For example:
````
PathPrefixAdapter::read($a, $b, $c) {
    $this->adapter->write('....');
    return $this->adapter->read($a, $b, $c);
}
````

Would not be detected in my tests, while it is obviously not good :|

### CODE REVIEW

File: src/Adapter/PathPrefixAdapter.php [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    {
+        return $this->adapter->readStream($this->applyPathPrefix($path));
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function listContents($directory = '', $recursive = false)
+    {
+        return $this->adapter->listContents($this->applyPathPrefix($directory));
```
Comment (frankdejonge) [inline src/Adapter/PathPrefixAdapter.php]: This needs to remove the path prefix afterwords as well to maintain transparency.

**Comment by frankdejonge** [comment]
@SamMousa what you can do is use mockery to create a mock of the AdapterInterface and express what you expect and what it should return. Alternatively you could just create a Stub for it which has pre-set responses.

Al in all, all the responses need to check for paths and remove the prefix from the path so the adapter has the same behaviour across the board.

**Comment by twistor** [comment]
Is `PathPrefixAdapter` actually useful? I'm not sure.

**Comment by SamMousa** [comment]
@twistor It allows you to specify a sort of namespace (path prefix) for any filesystem backend.
For example, if user X should only be able to do stuff in his own folder, I can use the path prefix adapter with a backend that does not natively support paths / path prefixes.

**Comment by twistor** [comment]
@SamMousa That makes sense I guess. I was thinking that most adapters their own way of handling prefixes, but I guess in theory this makes sense.

Can you rebase this pull request against master? We removed the HHVM tests.

**Comment by SamMousa** [comment]
It is on my todo list!

**Comment by SamMousa** [comment]
@twistor I have updated my PR.

**Pull request closed without merging**